### ISSUE TITLE
HEAD-handling code corrupts the HTTP channel in some cases, is generally redundant and confused in others

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3684]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3684|
|Type|defect|
|Created|2009-03-10 21:15:46Z|
|Branch|https://github.com/twisted/twisted/tree/head-3684|

If an UnsupportedMethod exception is thrown for rendering a HEAD method, the default for `Request.render` (if `GET` is in `allowedMethods`) is to call `render()` again with the request mutated to have its method be `GET`.  In some resources presumably this will return a body.

If the `Resource`'s `render_GET` returns `NOT_DONE_YET` (i.e. defers writing its body to the channel), the body will actually be returned to the client, violating the RFC.  If the method attribute were left as `HEAD`, then `http.Request.write` would enforce the no-data rule.

According to `trial --coverage twisted.web`, this code path is completely untested.  It's not clear that it's even necessary, given that `Resource.render_HEAD` delegates to `render_GET` by default.  (Or, vice versa.)

There's also more confused code at the end of `server.Request.render` that tries to specifically handle HEAD.

<details><summary>Searchable metadata</summary>

```
trac-id__3684 3684
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_head_3684 branches/head-3684
branch_author__washort washort
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1236719746000000 1236719746000000
changetime__1300306915000000 1300306915000000
version__None None
owner__washort washort

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to dash|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_2"></a>|@washort commented|
|-|-|

(In [28406]) Branching to head-3684

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_5"></a>|@wsanchez commented|
|-|-|

First, I like that your branch is called "head"

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_6"></a>|@wsanchez set owner to @washort<br>@wsanchez set status to `new`|
|-|-|

  OK, so with this API, if you have `allowedMethods = ["GET"]`, then even though you are telling us that you don't want to allow HEAD (for whatever reason), we're going to ignore you and allow it anyway.

  This strikes me as strangely inconsistent/exceptional API, so that bothers me.  What I would prefer is that if you say `allowedMethods = ["GET", "HEAD"]`, but you don't implement the `HEAD` method, then we'll use your `GET` implementation to provide `HEAD`.

  That is, we should do what you ask us to do, and not pretend that we know better.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez commented|
|-|-|

Otherwise, this patch looks fine

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_8"></a>|@washort commented|
|-|-|

It's always done this. [#4958](https://github.com/twisted/twisted/issues/4958) filed to address this problem.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_9"></a>|@washort set status to `closed`|
|-|-|

(In [31179]) 
Merge 'head-3684'

Author: washort
Reviewer: wsanchez
Fixes: [#3684](https://github.com/twisted/twisted/issues/3684)

When render_GET is called to satisfy HEAD requests, don't write a body in the response.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `reopened`|
|-|-|

(In [31184]) Revert r31179 - test suite regression

twisted.web.test.test_distrib.UserDirectoryTests.test_interface and
twisted.web.test.test_wsgi.WSGIResourceTests.test_interfaces fail the
verifyObject check because they do not provide the allowedMethods
attribute.

Reopens [#3684](https://github.com/twisted/twisted/issues/3684)

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_11"></a>|@washort set status to `closed`|
|-|-|

(In [31195]) Merge 'head-3684'

Author: washort
Reviewer: wsanchez
Fixes: [#3684](https://github.com/twisted/twisted/issues/3684)

When render_GET is called to satisfy HEAD requests, don't write a body in the response.
Remerge of r15826 which was reverted because IResource was incorrectly changed.
